File: Desktop/WinViewer/VB.NET/WinViewer/Helper/ViewerHelper.vb
Imports GrapeCity.ActiveReports.Viewer.Win.Internal.Export
Imports GrapeCity.ActiveReports.PageReportModel
Imports System.Xml.XPath
Imports System.Xml
Imports System.IO
Imports GrapeCity.ActiveReports

Namespace GrapeCity.ActiveReports.Viewer.Helper

	Public Class ViewerHelper
		''' <summary>
		''' Determines if the specified report is FPL report
		''' </summary>
		Public Shared Function DetermineReportType(fileName As FileInfo) As ExportForm.ReportType
			Dim extension As Object = Path.GetExtension(fileName.FullName).ToLowerInvariant()
			If extension = ".json" OrElse extension = ".bson" Then
				Return ExportForm.ReportType.PageCpl
			End If
			If extension = ".rdf" OrElse extension = ".rpx" Then
				Return ExportForm.ReportType.Section
			End If

			Dim report As PageReport
			Try
				report = New PageReport(fileName)
			Catch generatedExceptionName As ReportException
				Return ExportForm.ReportType.Section
			Catch generatedExceptionName As XmlException
				Return ExportForm.ReportType.Section
			End Try
			If IsNothing(report) OrElse IsNothing(report.Report) OrElse IsNothing(report.Report.Body) Then
				Return ExportForm.ReportType.Section
			End If

			Dim items As ReportItemCollection = report.Report.Body.ReportItems
			Return If(Not IsNothing(items) AndAlso items.Count = 1 AndAlso TypeOf items(0) Is FixedPage, ExportForm.ReportType.PageFpl, ExportForm.ReportType.PageCpl)
		End Function

		''' <summary>
		''' Checks file extansion.
		''' </summary>
		Public Shared Function IsRdf(fileName As FileInfo) As Boolean
			Dim extension As Object = Path.GetExtension(fileName.FullName)
			Return ".rdf".Equals(extension)
		End Function

		Public Shared Function GetReportServerUri(fileInfo As FileInfo) As String
			Dim content = File.ReadAllText(fileInfo.FullName)
			Dim document As XDocument = XDocument.Load(New StringReader(content))

			If IsNothing(document.Root) Then
				Return String.Empty
			End If

			Dim ns = document.Root.GetDefaultNamespace()
			Dim namespaceManager As New XmlNamespaceManager(New NameTable())
			namespaceManager.AddNamespace("ns", ns.NamespaceName)

			' try to get remote server uri from rdlx document
			Dim element = document.XPathSelectElement("/ns:Report/ns:CustomProperties/ns:CustomProperty[ns:Name[text()='ReportServerUri']]/ns:Value", namespaceManager)
			If Not IsNothing(element) Then
				Return element.Value
			End If

			' try to get remote server uri from rpx document
			element = document.XPathSelectElement("/ns:ActiveReportsLayout", namespaceManager)
			Dim attribute = If(Not IsNothing(element), element.Attribute("ReportServerUri"), Nothing)
			Return If(Not IsNothing(attribute), attribute.Value, String.Empty)
		End Function
	End Class
End Namespace

File: Desktop/WinViewer/C#/WinViewer/Helper/ViewerHelper.cs
using System;
using System.IO;
using System.Xml;
using System.Xml.Linq;
using System.Xml.XPath;
using GrapeCity.ActiveReports.PageReportModel;
using GrapeCity.ActiveReports.Viewer.Win.Internal.Export;

namespace GrapeCity.ActiveReports.Viewer.Helper

{
	internal static class ViewerHelper
	{
		/// <summary>
		/// Determines if the specified report is FPL report
		/// </summary>
		public static ExportForm.ReportType DetermineReportType(FileInfo fileName)
		{
			var extension = Path.GetExtension(fileName.FullName).ToLowerInvariant();
			if (extension == ".json" || extension == ".bson") return ExportForm.ReportType.PageCpl;
			if (extension == ".rdf" || extension == ".rpx") return ExportForm.ReportType.Section;

			PageReport report;
			try
			{
				report = new PageReport(fileName);
			}
			catch (ReportException)
			{
				return ExportForm.ReportType.Section;
			}
			catch (XmlException)
			{
				return ExportForm.ReportType.Section;
			}
			if (report == null || report.Report == null || report.Report.Body == null) return ExportForm.ReportType.Section;

			ReportItemCollection items = report.Report.Body.ReportItems;
			return items != null && items.Count == 1 && items[0] is FixedPage ? ExportForm.ReportType.PageFpl : ExportForm.ReportType.PageCpl;
		}

		/// <summary>
		/// Checks file extansion.
		/// </summary>
		public static bool IsRdf(FileInfo fileName)
		{
			var extension = Path.GetExtension(fileName.FullName);
			return ".rdf".Equals(extension, StringComparison.InvariantCultureIgnoreCase);
		}

		public static string GetReportServerUri(FileInfo file)
		{
			var content = File.ReadAllText(file.FullName);
			XDocument document = XDocument.Load(new StringReader(content));

			if (document.Root == null)
				return string.Empty;

			var ns = document.Root.GetDefaultNamespace();
			var namespaceManager = new XmlNamespaceManager(new NameTable());
			namespaceManager.AddNamespace("ns", ns.NamespaceName);

			// try to get remote server uri from rdlx document
			var element = document.XPathSelectElement("/ns:Report/ns:CustomProperties/ns:CustomProperty[ns:Name[text()='ReportServerUri']]/ns:Value", namespaceManager);
			if (element != null)
				return element.Value;

			// try to get remote server uri from rpx document
			element = document.XPathSelectElement("/ns:ActiveReportsLayout", namespaceManager);
			var attribute = element != null ? element.Attribute("ReportServerUri") : null;
			return attribute != null ? attribute.Value : string.Empty;
		}
	}
}
